index.d.ts

```diff
@@ -28,36 +28,64 @@ declare module "hyperapp" {
 
   // ---------------------------------------------------------------------------
 
-  // A Hyperapp instance has an initial state and a base view.
-  // It's usually mounted over an available DOM element.
-  type App<S> =
-    | Readonly<{ init: State<S> | StateWithEffects<S> | Action<S> }>
-    | Readonly<{ subscriptions: Subscriptions<S> }>
-    | Readonly<{ dispatch: DispatchInitializer<S> }>
-    | Readonly<{
-        init?: State<S> | StateWithEffects<S> | Action<S>
-        view: View<S>
-        node: Node
-        subscriptions?: Subscriptions<S>
-        dispatch?: DispatchInitializer<S>
-      }>
+  // This lets you use a version of `h` which assumes your particular app state.
+  interface TypedH<S, C = unknown> {
+    <_ extends never, T extends string = string>(
+      tag: T extends "" ? never : T,
+      props: PropList<S, C>,
+      children?: MaybeVDOM<S> | readonly MaybeVDOM<S>[]
+    ): VDOM<S>
+  }
+
+  // ---------------------------------------------------------------------------
+
+  // This utility type requires every property of an object or none at all.
+  // `App` uses this to make sure `view:` always appears alongside `node:`.
+  type AllOrNothing<T> = T | { [K in keyof T]?: never }
+
+  // This utility type ensures at least one property in an object to be present.
+  // `App` uses this to conveniently invalidate `app({})` because that is
+  // basically a fancy no-op.
+  type AtLeastOne<T> = { [K in keyof T]: Pick<T, K> }[keyof T]
+  // Credit: https://stackoverflow.com/a/59987826/1935675
+
+  // This utility type validates plain objects while invalidating array objects
+  // and string objects by disallowing numerical indexing. Used by `StyleProp`.
+  type IndexableByKey = Record<number, never>
+
+  // ---------------------------------------------------------------------------
+
+  // A Hyperapp instance generally has an initial state and a base view and is
+  // mounted over an available DOM element.
+  type App<S> = Readonly<
+    {
+      view: View<S>
+      node: Node
+    } |
+    AtLeastOne<{
+      init: Dispatchable<S>
+      subscriptions: Subscriptions<S>
+      dispatch: DispatchInitializer<S>
+    }> &
```

Comment:
Actually, you can create an app without `init`, `subscriptions`, or `dispatch `. Why is this necessary?

Reply:
`app({ view, node })` works but `app({})` has no redeeming value.

Reply:
`app({})` is not an error, so the types shouldn't disallow that.

Reply:
The types disallow many things that are not errors. For example, `text()`. Part of the role of types is to eliminate obviously useless code.  What does `app({})` do better than `() => {}` besides waste resources?

Also, consider scenarios where Hyperapp instances are dynamically generated.  If I have `app(generatedAppProps)` somewhere in my setup code I would like to know if there's a regression that's causing `generatedAppProps` to be empty for some reason while editing/compiling.

Reply:
Okay, maybe create an issue (or even better, start a discussion) and let's talk about it? 

Until then, let's not error on `app({})`?

Reply:
Sorry, I don't see the problem.  It's already, IMHO, an obscure edge case and I've mentioned two reasons to treat it as bad.

Reply:
If someone really really wanted a "no-op" `app()` call then `app({ init: void this })` is at least catchy.

Reply:
I want to defer making this decision after other things have been solved, like making the types more ergonomic and user-friendly. Right now, erroring on `app({})` seems is not a priority and I am not 100% sure I want to have this. If I accept this, I'll be unable to cut a new release until I make that decision.

Reply:
I want to defer this after other things have been worked on, like making the types more ergonomic and user-friendly. Right now, erroring on `app({})` is not a priority and I am not even 100% convinced we want to have this. If I accept this now, I'll be unable to cut a new release until I'm convinced.

Reply:
I want to defer this after other things have been worked on, like making the types more ergonomic and user-friendly. Right now, erroring on `app({})` is not a priority and I am not even 100% convinced we want this. If I accept this now, I'll be unable to cut a new release.

Reply:
I hear what you're saying but you're asking me to put in more effort to undo the protections I believe in without giving me an example of where `app({})` is the preferred choice and also without refuting the issues I've brought up about allowing it.

Besides, disallowing `app({})` increases the ergonomics of the embedded Hyperapp instance case because I can be rest assured that the types will let me know if my app configuration generation code is doing nothing.

Reply:
Also, it's not like we're talking about a complex situation here. Again, it seems like an open-and-shut case to me.

Reply:
And if there truly is somehow a use case for `app({})` then that can be addressed in the future.

Reply:
I want to take the reverse approach. Start simple, and if there's a necessity to forbid `app({})`, do it only then.

Reply:
Okay, but I already told you why it's beneficial to forbid it.

Reply:
Consider this scenario:

Let's say I'm working on a larger project with a team of developers and we have this relevant code in the project:

```ts
// farawaySetup.ts
const configureProperly = () => ({ init: 42 })
```

```ts
// embedded.ts
import { app } from "hyperapp"
import { configureProperly } from "../somewhere/else/farawaySetup.ts"
app(configureProperly())
```

So far, so good.

Now, let's say I pull in a future change from a new team member who committed this:

```ts
// farawaySetup.ts
const configureProperly = () => ({})
```

Obviously, that's contrived but I'm using that for sake of brevity. In a real-word scenario there would of course be more going on that would ultimately result in the same way.

Anyway, without having to check or know to check `farawaySetup.ts` **pre-emptively**, I'll know immediately that something went wrong somewhere during my colleague's commit process. For instance, was the githook that should have been setup to run a type/compile check on code being committed somehow get misconfigured?

This benefit is thanks to `app({})` being disallowed.

However, if `app({})` _were_ allowed then I wouldn't be privy to the issue until testing time or runtime. Therefore wasting time.